PanelEdit: Edit the source panel, refactor out VizPanelManager, simplify

torkelo opened the pull request:
* [x] Edit & render source panel. Radically simplifies and removes a lot of complex code in multiple locations 
* [x] Make the different components of PanelEditor more standalone (ie they no longer reference VizPanelManager but handle things that are relevant themselves). PanelOptionsPane handles changing panel plugin, PanelDataQueriesTab handles loading/changing data source, updating queries etc. Now these components could be easily reused outside the full screen panel edit (in the future if we want to).
* [x] Remove VizPanelManager. it serves no purpose and adds complexity, spread out the logic in it where it was called from
* [x] Discard restores original state 
* [x] Save from panel edit should reset original state (that is set when you discard). 
* [x] Repeat options work 
* [x] Repeat options that should display conditionally (when variable is setI)
* [x] Panel inspector works (now without any special panel edit logic)
* [x] Save work (now without any special panel edit logic)
* [x] Update PanelEditor tests
* [x] Move VizPanelManager tests to where the logic was moved
* [x] Update DataPlane tests 
* [x] Make library panel editing work 
* [x] Make library panel editing of repeat options work 
* [x] Unlink library panel code moved from VizPanelManager to PanelEditor and to LibraryPanelBehavior (And unit test for this that was on VizPanelManager is on PanelEditor now) 
* [x] Moved and cleanup a lot of tests from VizPanelManager.test to PanelDataQueriesTab.test.tsx (everything that deals with data source, query options, queries changes) 
* [x] Remove ResolveInspectPanelByKey no longer needed 

Future things maybe
* [ ] Remove buildEditPanelScene and replace with static factory function

torkelo committed:
Options pane, data pane queries tab and transformations tab working

torkelo committed:
Update

torkelo committed:
Discard works

torkelo committed:
Panel inspect working

torkelo committed:
Table viw works

torkelo committed:
Repeat options

torkelo committed:
Began fixing tests

torkelo committed:
More tests fixed

torkelo committed:
Progress on tests

torkelo committed:
no errors

torkelo committed:
Merge branch 'main' of github.com:grafana/grafana into panel-edit-major-refactor

torkelo committed:
init full width when enabling repeat

torkelo committed:
Began moving VizPanelManager tests to where the code was moved

torkelo committed:
Unlink libray panel code and unit test

torkelo committed:
Fixes and unit tests for change tracking and resetting original state and dirty flag when saving

torkelo committed:
migrating and improving unit tests

torkelo committed:
Done with VizPanelManager tests refactoring

torkelo commented on public/app/features/dashboard-scene/panel-edit/PanelDataPane/DataProviderSharer.tsx:
```diff
@@ -0,0 +1,48 @@
+  source: SceneObjectRef<SceneDataProvider>;
+}
+
+export class DataProviderSharer extends SceneObjectBase<DataProviderSharerState> implements SceneDataProvider {
```
this is a util data provider (that we maybe should add to scenes lib), that allow one scene object to share the data from another. Used by the table view to share the data from the original panel

torkelo committed:
Update

torkelo committed:
Update

torkelo committed:
remove console.log

torkelo committed:
Removed unnesssary behavior and fixed test

torkelo committed:
Update

torkelo committed:
Fix unrelated test

torkelo committed:
conditional options fix

torkelo committed:
remove

torkelo committed:
erge branch 'main' of github.com:grafana/grafana into panel-edit-major-refactor

torkelo commented:
The only thing not working exactly like before is when you edit a repeated panel, your editing the "template" panel, not one of the repeats so you get no local variable (scoped var), so the variable has all the values. So depending on your query this could lead to issues but it also feels more correct, your editing a panel that you are repeating based on a variable, so while in panel edit it will have all values. 

I see two options
* A. Keep as is in this PR, if the variable you repeat by has multiple values it will have multiple values in panel edit
* B. We could change the variable when you go into panel edit to only a single value (if multiple valued) and restore the value when leaving panel edit 

for panels inside repeated rows it just works as the panel being rendered is the panel inside the first row so has a local scoped variable. The problem is repeated panels inside DashboardGridItem where you edit state.body not the first state.repeatedPanels[0]

torkelo committed:
Fixing issue with editing repeated panels and scoping variable to first value

torkelo commented:
Just fixed the above mentioned issue by adding the same variable scope as the first repeated panel to the state.body VizPanel of a DashbordGridItem (the template panel)

torkelo committed:
Minor fix

kaydelaney commented:
Amazing work on this! Tested out the PR but it seems like discard panel changes doesn't seem to be working:

https://github.com/user-attachments/assets/e796f256-2a4d-4d3e-bb4f-6bb24452513f

torkelo committed:
Merge branch 'main' of github.com:grafana/grafana into panel-edit-major-refactor

kaydelaney reviewed (commented):
Looks good! Just some small nits, feel free to ignore.

torkelo committed:
Fix discard query runner changes

torkelo commented:
@kaydelaney thanks for the review! and for finding the discard issue, had not tested discarding query runner changes, only VizPanel level state. fixed (and added test for it).

torkelo committed:
Review comment changes

dprokop reviewed (commented):
Code looks good, a small nit comment only.

I found a bug tho. Try the attached dashboard. Edit second panel - change viz type, and discard changes. Runtime error occurs, seems like panel options change is buggy with this PR.


<details>
<summary>dash json</summary>

```json
{
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": {
          "type": "grafana",
          "uid": "-- Grafana --"
        },
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "fiscalYearStartMonth": 0,
  "graphTooltip": 0,
  "id": 2512,
  "links": [],
  "panels": [
    {
      "datasource": {
        "type": "prometheus",
        "uid": "fdloj1mdh4jcwe"
      },
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "axisBorderShow": false,
            "axisCenteredZero": false,
            "axisColorMode": "text",
            "axisLabel": "",
            "axisPlacement": "auto",
            "barAlignment": 0,
            "barWidthFactor": 0.6,
            "drawStyle": "line",
            "fillOpacity": 0,
            "gradientMode": "none",
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "insertNulls": false,
            "lineInterpolation": "linear",
            "lineWidth": 1,
            "pointSize": 5,
            "scaleDistribution": {
              "type": "linear"
            },
            "showPoints": "auto",
            "spanNulls": false,
            "stacking": {
              "group": "A",
              "mode": "none"
            },
            "thresholdsStyle": {
              "mode": "off"
            }
          },
          "links": [
            {
              "title": "Filter",
              "url": "http://localhost:3000/d/bdmh0z3p3pf5sa/filters-demo?var-query0=code%7C%3D%7C${__field.labels.code}"
            }
          ],
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 16,
        "w": 12,
        "x": 0,
        "y": 0
      },
      "id": 2,
      "options": {
        "legend": {
          "calcs": [],
          "displayMode": "list",
          "placement": "bottom",
          "showLegend": true
        },
        "tooltip": {
          "mode": "single",
          "sort": "none"
        }
      },
      "pluginVersion": "11.3.0-pre",
      "targets": [
        {
          "datasource": {
            "type": "prometheus",
            "uid": "fdloj1mdh4jcwe"
          },
          "disableTextWrap": false,
          "editorMode": "builder",
          "expr": "rate(prometheus_http_requests_total{code=~\"500|503|400\"}[$__rate_interval])",
          "fullMetaSearch": false,
          "hide": false,
          "includeNullMetadata": false,
          "instant": false,
          "legendFormat": "__auto",
          "range": true,
          "refId": "A",
          "useBackend": false
        }
      ],
      "title": "filter applicable",
      "type": "timeseries"
    },
    {
      "datasource": {
        "type": "prometheus",
        "uid": "fdloj1mdh4jcwe"
      },
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "thresholds"
          },
          "links": [
            {
              "title": "Filter",
              "url": "http://localhost:3000/d/bdmh0z3p3pf5sa/filters-demo?var-query0=code%7C%3D%7C${__field.labels.code}"
            }
          ],
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 16,
        "w": 12,
        "x": 12,
        "y": 0
      },
      "id": 3,
      "options": {
        "colorMode": "value",
        "graphMode": "area",
        "justifyMode": "auto",
        "orientation": "auto",
        "percentChangeColorMode": "standard",
        "reduceOptions": {
          "calcs": [
            "lastNotNull"
          ],
          "fields": "",
          "values": false
        },
        "showPercentChange": false,
        "textMode": "auto",
        "wideLayout": true
      },
      "pluginVersion": "11.3.0-pre",
      "targets": [
        {
          "datasource": {
            "type": "prometheus",
            "uid": "fdloj1mdh4jcwe"
          },
          "disableTextWrap": false,
          "editorMode": "builder",
          "expr": "rate(prometheus_http_requests_total{code=~\"500|503|400\"}[$__rate_interval])",
          "fullMetaSearch": false,
          "hide": false,
          "includeNullMetadata": false,
          "instant": false,
          "legendFormat": "__auto",
          "range": true,
          "refId": "A",
          "useBackend": false
        }
      ],
      "title": "filter applicable",
      "type": "stat"
    }
  ],
  "preload": false,
  "schemaVersion": 39,
  "tags": [],
  "templating": {
    "list": [
      {
        "current": {
          "text": "",
          "value": ""
        },
        "name": "query2",
        "options": [],
        "query": "",
        "refresh": 1,
        "regex": "",
        "type": "query"
      },
      {
        "current": {
          "text": [
            ""
          ],
          "value": [
            ""
          ]
        },
        "name": "GroupBy",
        "type": "groupby"
      },
      {
        "baseFilters": [],
        "datasource": {
          "type": "prometheus",
          "uid": "fdloj1mdh4jcwe"
        },
        "filters": [
          {
            "condition": "",
            "key": "handler",
            "keyLabel": "handler",
            "operator": "!=|",
            "value": "/-/quit",
            "valueLabels": [
              "/-/quit",
              "/-/ready",
              "/-/reload",
              "/alerting/",
              "/alerts",
              "/admin",
              "/api/alert-notifiers",
              "/alerting/*",
              "/api/datasources/proxy/uid/:uid/*",
              "/api/ds/query",
              "/api/health",
              "/api/user/preferences",
              "/api/short-urls",
              "/api/v1/series",
              "/api/frontend/settings/",
              "/-/healthy",
              "/api/frontend/assets",
              "/dashboards/",
              "/consoles/*filepath",
              "/api/v1/admin/tsdb/clean_tombstones",
              "/api/folders/",
              "/favicon.ico",
              "/api/user/password/send-reset-email"
            ],
            "values": [
              "/-/quit",
              "/-/ready",
              "/-/reload",
              "/alerting/",
              "/alerts",
              "/admin",
              "/api/alert-notifiers",
              "/alerting/*",
              "/api/datasources/proxy/uid/:uid/*",
              "/api/ds/query",
              "/api/health",
              "/api/user/preferences",
              "/api/short-urls",
              "/api/v1/series",
              "/api/frontend/settings/",
              "/-/healthy",
              "/api/frontend/assets",
              "/dashboards/",
              "/consoles/*filepath",
              "/api/v1/admin/tsdb/clean_tombstones",
              "/api/folders/",
              "/favicon.ico",
              "/api/user/password/send-reset-email"
            ]
          }
        ],
        "hide": 1,
        "name": "Filter",
        "type": "adhoc"
      }
    ]
  },
  "time": {
    "from": "now-6h",
    "to": "now"
  },
  "timepicker": {
    "hidden": true
  },
  "timezone": "browser",
  "title": "Filters demo",
  "uid": "bdmh0z3p3pf5sa",
  "version": 22,
  "weekStart": ""
}
```
</details>

```
An unexpected error happened
Details
TypeError: Cannot read properties of undefined (reading 'cellOptions')

    at http://localhost:3000/public/build/app.b67f6d3decf2593a14b5.js:237083:65
    at div
    at push../public/app/features/dashboard-scene/panel-edit/PanelOptionsPane.tsx.Component (http://localhost:3000/public/build/app.b67f6d3decf2593a14b5.js:237237:25)
    at SceneComponentWrapperWithoutMemo (http://localhost:3000/public/build/app.b67f6d3decf2593a14b5.js:33789:18)
    at div
    at div
    at PanelEditorRenderer (http://localhost:3000/public/build/app.b67f6d3decf2593a14b5.js:236856:32)
    at SceneComponentWrapperWithoutMemo (http://localhost:3000/public/build/app.b67f6d3decf2593a14b5.js:33789:18)
    at div
    at Page (http://localhost:3000/public/build/app.b67f6d3decf2593a14b5.js:199492:3)
    at DashboardSceneRenderer 
```

https://github.com/user-attachments/assets/d52f97b3-cb6c-4479-8139-5732b1c09b4d

bfmatei reviewed (commented):
I didn't get the chance to test this thoroughly, but the code itself looks good.

torkelo commented:
> Testing this more, seems like discard doesn't work in many cases. I.e. add panel -> change panel title -> discard -> runtime error

@dprokop unable to replicate, discard always works for me. strange, is there something else on the dashboard you have?

torkelo commented:
@dprokop ah, I missed you had a new panel. I can replicate it there. 

will fix, something wrong with the isNewPanel logic

torkelo committed:
fix discard issue with new panels

torkelo committed:
Add loading state to panel edit

torkelo commented:
@dprokop Pushed another update to make discard even more rebust for some new edge cases. 

To solve those I added a new loading (initalization) phase to PanelEditor so that it waits for the panel plugin to load before initializing options pane and setting the original state discard will discard to. This is needed for full page reload of panels that auto migrate from angular to react or just panels with option migrations. The trick is for PanelEditor to wait until plugin is loaded before it set the state discard will discard to.

this waiting for plugin to load before initializing is great and was something I wanted to add anyway as it will help going into panel edit directly (full pag reload) for lib panels, we can remove the wait for lib panel to load in DashboardUrlSync (but will do that in a a follow-up PR).

torkelo commented:
added a new unit tests to PanelEditor for the waiting / init phase and made all those test async so they wait for the plugin to load

torkelo committed:
Fix test

torkelo committed:
Merge branch 'main' of github.com:grafana/grafana into panel-edit-major-refactor

torkelo committed:
Update

torkelo committed:
fix test

torkelo committed:
fix lint

torkelo committed:
lint

torkelo commented on public/app/features/dashboard-scene/panel-edit/PanelDataPane/DataProviderSharer.tsx:
```diff
@@ -0,0 +1,48 @@
+  source: SceneObjectRef<SceneDataProvider>;
+}
+
+export class DataProviderSharer extends SceneObjectBase<DataProviderSharerState> implements SceneDataProvider {
```
@dprokop https://github.com/grafana/scenes/pull/903

torkelo committed:
Fix

torkelo committed:
Fix overrides editing  mutating fieldConfig

torkelo committed:
Merge branch 'main' of github.com:grafana/grafana into panel-edit-major-refactor

axelavargas committed:
Merge remote-tracking branch 'origin' into panel-edit-major-refactor

mdvictor commented:
I think there is an issue with repeated panels here. If you edit a repeated panel and go back to the dashboard the changes aren't saved:

https://github.com/user-attachments/assets/36bde1cb-8550-4929-8166-0e9bd6e8a226

kaydelaney reviewed (approved):
Amazing work! Looks good, and couldn't find any issues with manual testing

kaydelaney reviewed (changes requested):
Oops, looks like I approved too early - seems like Oscar found some issues

oscarkilhed reviewed (commented):
https://github.com/user-attachments/assets/e51f48fa-5c4d-4f38-b58a-9dc340154bba


https://github.com/user-attachments/assets/2ff284cc-020b-4638-93e9-fa0d7fb3617b

Found these issues where stat and gauge panel does not update for all options. The update is visible once some other option changes.

kaydelaney commented:
Checked the issues @oscarkilhed pointed out against main and I was able to reproduce both, so I don't think they're caused by this PR.

kaydelaney reviewed (approved):
Reapproving! LGTM!
